drivers/mcp2515/candev_mcp2515.c

```diff
@@ -449,66 +451,67 @@ static int _set_filter(candev_t *dev, const struct can_filter *filter)
         f.can_mask &= CAN_SFF_MASK;
     }
 
-    /* Browse on each mailbox to find an empty space */
-    int mailbox_index = 0;
-
-    while (mailbox_index < MCP2515_RX_MAILBOXES) {
-        /* mask unused */
-        if (dev_mcp->masks[mailbox_index] == 0) {
-            /* set mask */
-            mcp2515_set_mask(dev_mcp, mailbox_index, f.can_mask);
-            /* set filter */
-            mcp2515_set_filter(dev_mcp, MCP2515_FILTERS_MB0 * mailbox_index,
-                               f.can_id);
+    /* mask unused */
+    if (dev_mcp->masks[f.target_mailbox] == 0) {
+        /* set mask */
+        mcp2515_set_mask(dev_mcp, f.target_mailbox, f.can_mask);
+        /* set filter */
+        mcp2515_set_filter(dev_mcp, MCP2515_FILTERS_MB0 * f.target_mailbox,
+                           f.can_id);
 
-            /* save filter */
-            dev_mcp->masks[mailbox_index] = f.can_mask;
-            dev_mcp->filter_ids[mailbox_index][0] = f.can_id;
+        /* save filter */
+        dev_mcp->masks[f.target_mailbox] = f.can_mask;
+        dev_mcp->filter_ids[f.target_mailbox][0] = f.can_id;
+    }
 
-            /* function succeeded */
-            break;
+    /* mask existed and same mask */
+    else if (dev_mcp->masks[f.target_mailbox] == f.can_mask) {
+        /* find an empty space if it exists */
+        int filter_pos = 1; /* first one is already filled */
+        /* stop at the end of mailbox or an empty space found */
+        while (filter_pos < _max_filters(f.target_mailbox) &&
+               dev_mcp->filter_ids[f.target_mailbox][filter_pos] != 0) {
+            filter_pos++;
         }
 
-        /* mask existed and same mask */
-        else if (dev_mcp->masks[mailbox_index] == f.can_mask) {
-            /* find en empty space if it exist */
-            int filter_pos = 1; /* first one is already filled */
-            /* stop at the end of mailbox or an empty space found */
-            while (filter_pos < _max_filters(mailbox_index) &&
-                   dev_mcp->filter_ids[mailbox_index][filter_pos] != 0) {
-                filter_pos++;
+        /* an empty space is found */
+        if (filter_pos < _max_filters(f.target_mailbox)) {
+            /* set filter on this memory space */
+            if (mutex_trylock(&_mcp_mutex)) {
```

Comment:
Can this be called from Interrupt context? Why not just block here?

Reply:
The approach taken concerning the mutex (when writing the driver) seems to be opting for error return instead of blocking. I don't know whether this might be related to a CAN spec to not make a node connected to the CAN bus in blocked state or other reasons.